theano/tensor/signal/downsample.py

```diff
@@ -518,6 +600,15 @@ def __init__(self, ds, ignore_border, st=None, padding=(0, 0), mode='max'):
                 " 'average_inc_pad' and 'average_exc_pad'. Got %s" % mode)
         self.mode = mode
 
+    def infer_shape(self, node, in_shapes):
+        return [in_shapes[0]]
+
+
+class MaxPoolGrad(PoolGrad):
```

Comment:
As you aren't sure, I would go with my suggestion :)

@sebastien-j for optimization, the doc is there:

https://github.com/abergeron/ccw_tutorial_theano

in the advanced.pdf file.

pseudo code:

```
@register_canonicalize
@gof.local_optimizer([MaxPoolGrad])
def local_average_pool_grad(node):
    if not isinstance(node.op, T.MaxPoolGrad) or node.op.mode not in ['average_inc*' , 'aevrage_ex*']:
        return False
    return [AveragePoolGrad(ds=node.op.ds, ignore_border=node.op.ignore_border, st=node.op.st, padding=node.op.padding, mode=node.op.mode)(node.inputs[0], node.inputs[2])]
```